sycl/doc/design/GroupSort.md

```diff
@@ -0,0 +1,220 @@
+# Group sort algorithm
+
+Group sorting algorithms are needed to sort data without calling additional kernels
+They are described by SYCL 2020 Extension specification:
+[direct link to the specification's extension][group_sort_spec].
+
+[group_sort_spec]: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc
+
+Example usage:
+
+```cpp
+#include <sycl/sycl.hpp>
+
+namespace oneapi_exp = sycl::ext::oneapi::experimental;
+sycl::range<1> local_range{256};
+// predefine radix_sorter to calculate local memory size
+using RSorter = oneapi_exp::radix_sorter<T, oneapi_exp::sorting_order::descending>;
+// calculate required local memory size
+size_t temp_memory_size =
+    RSorter::memory_required(sycl::memory_scope::work_group, local_range);
+q.submit([&](sycl::handler& h) {
+  auto acc = sycl::accessor(buf, h);
+  auto scratch = sycl::local_accessor<std::byte, 1>( {temp_memory_size}, h);
+  h.parallel_for(
+    sycl::nd_range<1>{ local_range, local_range },
+    [=](sycl::nd_item<1> id) {
+      acc[id.get_local_id()] =
+        oneapi_exp::sort_over_group(
+          id.get_group(),
+          acc[id.get_local_id()],
+          RSorter(sycl::span{scratch.get_pointer(), temp_memory_size})
+      );
+    });
+  });
+...
+```
+
+## Design objectives
+
+In DPC++ Headers/DPC++ RT we don't know which sorting algorithm is better for
+different architectures. Backends have more capability to optimize the sorting algorithm
+using low-level instructions.
+
+Data types that should be supported by backends: arithmetic types
+(https://en.cppreference.com/w/c/language/arithmetic_types), `sycl::half`.
+
+Comparators that should be supported by backends: `std::less`, `std::greater`,
+custom comparators
+
+## Design
+
+Overall, for backend support we need to have the following:
+- Fallback implementation of sorting algorithms for user's types, comparators and/or sorters.
+
+- Backend implementation for types, comparators and/or sorters
+  that can be optimized using backend specific instructions.
+
+  **NOTE**: It was decided that `radix_sorter` will be implemented only in DPC++ Headers since
+it's difficult to support such algorithm at backends' level.
+
+- Fallback implementation in case if backends don't have more optimized implementations yet.
+
+- Level Zero extension for `memory_required` functions.
```

Comment:
We do not directly call into Level Zero from SYCL RT, but instead we go through [unified runtime](https://github.com/oneapi-src/unified-runtime/tree/main). Therefore, if we need a new API to do queries to low-level runtimes, then UR should also be updated